File: apps/web/lib/validation/timeline.ts
/**
 * Zod schemas for the Timeline.
 *
 * API responses are external input too, so they are parsed before entering
 * application state — a backend change that alters the payload fails here,
 * loudly, instead of surfacing as `undefined` in an entry.
 *
 * There are **no form schemas**: the timeline is read-only. Events are published
 * by the services that cause them, and there is no endpoint through which a
 * client could write one.
 *
 * The rules deliberately mirror `apps/api/schemas/timeline.py`. Note where they
 * are *tolerant* rather than strict — `event_type` is a plain string and
 * `metadata` an open record — because the API's registry is an open set and a
 * strict enum here would turn a valid response from a newer backend into a client
 * error. `category` is the exception: the server computes it and can only ever
 * send one of five values.
 */

import { z } from "zod";

import { SORT_ORDERS } from "@/types/case-management";
import { TIMELINE_CATEGORIES } from "@/types/timeline";
import { TIMELINE_SORT_FIELDS } from "@/types/timeline-management";

/** Matches `MAX_EVENT_TYPE_LENGTH` in `apps/api/schemas/timeline.py`. */
export const MAX_EVENT_TYPE_LENGTH = 60;

/** A calendar date as `<input type="date">` produces it, and as the API expects it. */
const ISO_DATE = /^\d{4}-\d{2}-\d{2}$/;

// --------------------------------------------------------------------------- //
// Response schemas
// --------------------------------------------------------------------------- //

export const timelineEventSchema = z.object({
  id: z.string(),
  case_id: z.string(),

  event_type: z.string(),
  // Computed server-side from the event type, and defaulted there for anything
  // unrecognised — so an unknown value here would mean the *contract* changed,
  // not that a new event type shipped.
  category: z.enum(TIMELINE_CATEGORIES).catch("case"),

  title: z.string(),
  description: z.string().nullable().default(null),

  actor_id: z.string().nullable().default(null),
  actor_name: z.string().nullable().default(null),
  actor_role: z.string().nullable().default(null),

  // Open by design: the shape depends on the event type, and a future module
  // attaches its own keys without a client release.
  metadata: z.record(z.string(), z.unknown()).default({}),

  created_at: z.string(),
});

export const timelinePageSchema = z.object({
  items: z.array(timelineEventSchema),
  total_records: z.number().int().nonnegative(),
  page: z.number().int().positive(),
  page_size: z.number().int().positive(),
  total_pages: z.number().int().nonnegative(),
});

// --------------------------------------------------------------------------- //
// Query validation
// --------------------------------------------------------------------------- //

/**
 * The timeline query, validated before it becomes a query string.
 *
 * The values come from UI state rather than from a user typing a URL, but they
 * are still the input to a request — and validating here means an out-of-range
 * page from a stale link is corrected rather than sent to the API to be rejected.
 */
export const timelineQuerySchema = z.object({
  page: z.coerce.number().int().min(1).catch(1),
  pageSize: z.coerce.number().int().min(1).max(100).catch(20),
  search: z.string().trim().max(200).catch(""),
  eventType: z.string().trim().max(MAX_EVENT_TYPE_LENGTH).nullable().catch(null),
  actorId: z.string().nullable().catch(null),
  dateFrom: z.string().regex(ISO_DATE).or(z.literal("")).catch(""),
  dateTo: z.string().regex(ISO_DATE).or(z.literal("")).catch(""),
  sortBy: z.enum(TIMELINE_SORT_FIELDS).catch("created_at"),
  sortOrder: z.enum(SORT_ORDERS).catch("desc"),
});

File: apps/api/schemas/timeline.py
"""Timeline request and response schemas.

Two responsibilities, the same two the case and document schemas carry:

* **Validation** — the list query's page bounds, filters, search term, and date
  coherence are enforced here, so routes stay thin and every rejection comes back
  in the standard envelope with a per-field message.
* **Serialization** — an event is returned with its category computed and its
  actor already denormalised onto the row, so a client renders an entry without a
  second lookup.

There is deliberately **no request schema for creating an event**. ``08-timeline.md``
specifies two read endpoints and nothing else: events are published by the
business services that cause them (see :class:`~services.timeline.TimelineService`),
never by a client. An audit trail a client can write directly is not an audit
trail.
"""

from __future__ import annotations

import math
import uuid
from datetime import date, datetime
from enum import StrEnum
from typing import Any, Self

from pydantic import AliasChoices, BaseModel, ConfigDict, Field, computed_field, field_validator, model_validator

from core.timeline import category_for, known_event_types
from models.timeline import TimelineEventCategory
from schemas.case import SortOrder

#: Default and maximum page sizes for the timeline. The ceiling exists so a
#: single request cannot be used to dump a case's entire history.
DEFAULT_PAGE_SIZE = 20
MAX_PAGE_SIZE = 100

#: Matches ``timeline_events.event_type``.
MAX_EVENT_TYPE_LENGTH = 60

#: Re-exported so the timeline API's sort direction is literally the same type as
#: the case and document APIs', rather than a third enum that happens to agree.
__all__ = [
    "DEFAULT_PAGE_SIZE",
    "MAX_EVENT_TYPE_LENGTH",
    "MAX_PAGE_SIZE",
    "SortOrder",
    "TimelineEventPage",
    "TimelineEventRead",
    "TimelineListQuery",
    "TimelineSortField",
]


# --------------------------------------------------------------------------- #
# Responses
# --------------------------------------------------------------------------- #


class TimelineEventRead(BaseModel):
    """One timeline event as returned to authorized clients.

    Carries exactly the fields ``08-timeline.md`` lists, plus a **computed**
    ``category`` so the client can pick an icon without a second copy of the
    event-type → category mapping.
    """

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    id: uuid.UUID = Field(description="Unique event identifier.")
    case_id: uuid.UUID = Field(description="Case this event belongs to.")

    event_type: str = Field(
        description=(
            "What happened, as a registry identifier. The platform currently records: "
            f"{', '.join(f'`{value}`' for value in known_event_types())}. Treat this as an "
            "open set — a later module may publish a type not listed here."
        )
    )
    title: str = Field(description="Short headline, e.g. `Document Uploaded`.")
    description: str | None = Field(
        default=None, description="Human sentence describing what happened."
    )

    actor_id: uuid.UUID | None = Field(
        default=None,
        description="Identifier of the user who acted, if the account still exists.",
    )
    actor_name: str | None = Field(
        default=None,
        description=(
            "The actor's display name **at the time of the event**. A snapshot, not a lookup: "
            "renaming a user must not rewrite history."
        ),
    )
    actor_role: str | None = Field(
        default=None, description="The actor's platform role at the time of the event."
    )

    metadata: dict[str, Any] = Field(
        default_factory=dict,
        # The ORM attribute is `event_metadata`, because `metadata` is
        # SQLAlchemy's own attribute on the declarative base. The wire field is
        # `metadata`, as the spec specifies, and both names validate.
        validation_alias=AliasChoices("event_metadata", "metadata"),
        description=(
            "Structured specifics of this event, e.g. `{\"from\": \"open\", \"to\": "
            "\"in_progress\"}`. Shape depends on the event type; always an object, never null."
        ),
    )

    created_at: datetime = Field(description="When the event happened.")

    @computed_field(  # type: ignore[prop-decorator]
        description=(
            "Family the event belongs to — `case`, `status`, `priority`, `assignment`, or "
            "`document`. Derived from the event type rather than stored, so it cannot drift; "
            "an unrecognised type is presented as `case`."
        ),
    )
    @property
    def category(self) -> TimelineEventCategory:
        """Computed, so the client's icon choice cannot disagree with the server."""
        return category_for(self.event_type)


class TimelineEventPage(BaseModel):
    """One page of a timeline.

    Carries the totals the spec requires so a client can render pagination
    controls without a second count query.
    """

    items: list[TimelineEventRead] = Field(description="Events on this page.")
    total_records: int = Field(description="Total events matching the filters, across all pages.")
    page: int = Field(description="Current page number (1-based).")
    page_size: int = Field(description="Maximum number of events per page.")
    total_pages: int = Field(description="Number of pages available for the current filters.")

    @classmethod
    def build(
        cls, items: list[TimelineEventRead], *, total: int, page: int, page_size: int
    ) -> TimelineEventPage:
        """Assemble a page, deriving ``total_pages`` from the total and size.

        An empty result still reports one page, so a client never renders
        "page 1 of 0".
        """
        return cls(
            items=items,
            total_records=total,
            page=page,
            page_size=page_size,
            total_pages=max(1, math.ceil(total / page_size)) if page_size else 1,
        )


# --------------------------------------------------------------------------- #
# Listing
# --------------------------------------------------------------------------- #


class TimelineSortField(StrEnum):
    """Sortable columns of the timeline.

    One member, because the spec names one: *Event Date*. Declared as an enum
    rather than hard-coded so the parameter is self-documenting in OpenAPI and so
    a second sort key is an addition rather than a signature change.
    """

    #: "Event Date" — when the event happened.
    CREATED_AT = "created_at"


class TimelineListQuery(BaseModel):
    """Validated query parameters for the timeline endpoints.

    Every filter is optional and they combine with AND, which is what the spec's
    "filters should be combinable" requires.
    """

    model_config = ConfigDict(extra="forbid")

    page: int = Field(default=1, ge=1, description="Page number, 1-based.")
    page_size: int = Field(
        default=DEFAULT_PAGE_SIZE,
        ge=1,
        le=MAX_PAGE_SIZE,
        description=f"Events per page (max {MAX_PAGE_SIZE}).",
    )
    search: str | None = Field(
        default=None,
        max_length=200,
        description="Case-insensitive match against the event title or description.",
    )
    event_type: str | None = Field(
        default=None,
        max_length=MAX_EVENT_TYPE_LENGTH,
        description=(
            "Only events of this type. A free identifier rather than a closed enum, so a type "
            "published by a later module stays filterable."
        ),
    )
    actor_id: uuid.UUID | None = Field(
        default=None, description="Only events recorded for this user."
    )
    date_from: date | None = Field(
        default=None, description="Only events that happened on or after this date."
    )
    date_to: date | None = Field(
        default=None, description="Only events that happened on or before this date."
    )
    sort_by: TimelineSortField = Field(
        default=TimelineSortField.CREATED_AT, description="Column to sort by."
    )
    sort_order: SortOrder = Field(
        default=SortOrder.DESC,
        description=(
            "Sort direction. Defaults to descending, so a timeline reads newest first — which "
            "is the reverse-chronological order the spec's display requires."
        ),
    )

    @field_validator("search")
    @classmethod
    def _normalize_search(cls, value: str | None) -> str | None:
        if value is None:
            return None
        return value.strip() or None

    @field_validator("event_type")
    @classmethod
    def _normalize_event_type(cls, value: str | None) -> str | None:
        # Identifiers are stored lowercase, so a filter typed in any case matches
        # rather than silently returning nothing.
        if value is None:
            return None
        return value.strip().lower() or None

    @model_validator(mode="after")
    def _check_range(self) -> Self:
        # An inverted range matches nothing, which reads as "the filter is
        # broken" rather than as the input error it is.
        if self.date_from is not None and self.date_to is not None and self.date_to < self.date_from:
            raise ValueError("`date_to` cannot be before `date_from`.")
        return self

    @property
    def offset(self) -> int:
        """Rows to skip for the requested page."""
        return (self.page - 1) * self.page_size
